---
title: "Bioinformatics in R. Lab 2: Data Visualization and GO analysis"
author: "J. Cesar Ignacio Espinoza - Cesar   "
date: "Week 03: April 1st and 3rd 2024"
output: 
  html_document: 
    highlight: espresso
    theme: cerulean
---

```{r}
# Install BiocManager if not already installed
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# Install DESeq2
BiocManager::install("DESeq2")
```

```{r setup}
    ### Edit this line so your notebook renders properly
    knitr::opts_knit$set(root.dir = normalizePath("~/Downloads/R")) 
```

## Differential Gene Expression From Counts Table to DeSeq2
Begin by downloading the data from (Same link as in today's canvas page), download to your working directory.

#Install the required librries and load the data.

You will have to install enhanced volcano, if this gives issues like last time, we can try a different software.
```{r}
BiocManager::install("EnhancedVolcano")
```

We now load the libraries
```{r message=FALSE, warning=FALSE, paged.print=FALSE, results="hide"}
# We first need to import the important libnrary for today's class, dplyr
library(dplyr)
library(readr)
library(DESeq2)
library(ggplot2)
library(EnhancedVolcano)

```

Here we will run the entire dseq program to find significantly expressed genes

```{r}
### Run this chunk to import the counts table and metadata into your evironment.
counts <- read.csv('airway_scaledcounts.csv', row.names = 1, header = TRUE)
metadata <- read.csv('airway_metadata.csv', row.names = 1, header = TRUE)
### Modify the metadata table
sample_info <- metadata %>%
  dplyr::select(-geo_id) %>%
  dplyr::rename(dexamethasone = dex)
sample_info

#creat a deseqobject

dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData = sample_info,
                              design = ~ dexamethasone)
## Subset your ddseq object
subset <-  dds[(rowSums(counts(dds)) >= 10),]

### fix the base to a preferred treatment
relevel(dds$dexamethasone, ref='control')

### Run deseq2

deseq_ob <- DESeq(subset)

#### Save the results to a new object
res <- results(deseq_ob, alpha = 0.05)
```

### New packages: 
This package will be able to convert between ENSEMBL ID and symbol (gene names) 

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("org.Hs.eg.db")
```

```{r}
library("org.Hs.eg.db")
sigs.df <-  as.data.frame(res)
sigs.df$symbol <- mapIds(org.Hs.eg.db, keys= rownames(sigs.df), keytype = 'ENSEMBL', colum = "SYMBOL")
```

```{r}
sigs.df
```
### Make your first volcano plot!
```{r}
EnhancedVolcano(sigs.df, x='log2FoldChange', y = 'padj', lab = sigs.df$symbol)

```
Look at the manual (vignette for enhanced volcano) https://bioconductor.org/packages/devel/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano.html
Create a new code chunk and plot a better version of this volcano plot! (15 minutes)

```{r}
# Improved Volcano Plot
EnhancedVolcano(sigs.df, 
                x = 'log2FoldChange', 
                y = 'padj',
                lab = sigs.df$symbol,
                title = 'Dexamethasone Treatment vs Control',
                subtitle = 'Differential Gene Expression',
                pCutoff = 0.05,
                FCcutoff = 1.0,
                pointSize = 3.0,
                labSize = 4.0,
                colAlpha = 0.8,
                legendPosition = 'right',
                legendLabSize = 10,
                legendIconSize = 4.0,
                drawConnectors = TRUE,
                widthConnectors = 0.5,
                colConnectors = 'grey50',
                maxoverlapsConnectors = 25)
```

## Heatmaps!
```{R}
# Install BiocManager if not already available
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# Install ComplexHeatmap
BiocManager::install("ComplexHeatmap")
library(ComplexHeatmap)
```

```{R}
diff.df <- as.data.frame(sigs.df)
diff.df <- diff.df %>%
  filter(padj < 0.05)
```

```{R}
mat <- counts(deseq_ob, normalized = T)[rownames(diff.df),]
mat.z <- t(apply(mat,1, scale))
colnames(mat.z) <- colnames(mat)
Heatmap(mat.z, cluster_rows= T, cluster_columns= T, name = "Z-score", row_labels = diff.df[rownames(mat.z),]$symbol)
```
## Use dplyr to subset the table of significant values until you get a nice looking figure. 
(You cannot plot all genes at once so you have to subset in order to gain some insights from 
your data)
```{R}
diff.df <- as.data.frame(sigs.df)
diff.df <- diff.df %>%
  filter(padj < 0.05, log2FoldChange > 3.5, baseMean > 100)

mat <- counts(deseq_ob, normalized = T)[rownames(diff.df),]
mat.z <- t(apply(mat,1, scale))
colnames(mat.z) <- colnames(mat)
Heatmap(mat.z, cluster_rows= T, cluster_columns= T, name = "Z-score", row_labels = diff.df[rownames(mat.z),]$symbol)

```
## Gene Ontology Analysis
```{R}
BiocManager::install('clusterProfiler')
library(clusterProfiler)
```

```{R}
diff.df <- as.data.frame(sigs.df)
diff.df <- diff.df %>%
  filter(padj < 0.05, log2FoldChange < 1)

genes_to_test <- rownames(diff.df)
GO_results <- enrichGO(gene = genes_to_test, OrgDb = "org.Hs.eg.db", keyType = 'ENSEMBL', ont = "BP")
fit <- plot(barplot(GO_results, showCategory=10))
```

Anothe program is gost, we can try it below.
```{R}
library(gprofiler2) 
go_upreg <- gost(genes_to_test, organism = 'hsapiens', significant = T)
gostplot(go_upreg, interactive = T)
```
It might be more informative to create plots for genes that are both upregulated and downregulated. Subset your data above and plot the GO enrichment for both upregulated and downreghulated.
```{R}
# Code for upregulated and downregulated GO analysis
# First, identify upregulated and downregulated genes
diff.df <- as.data.frame(sigs.df)
upreg.df <- diff.df %>%
  filter(padj < 0.05, log2FoldChange > 0)

downreg.df <- diff.df %>%
  filter(padj < 0.05, log2FoldChange < 0)

# Get gene IDs for upregulated genes
upreg_genes <- rownames(upreg.df)
upGO_results <- enrichGO(gene = upreg_genes, 
                         OrgDb = "org.Hs.eg.db", 
                         keyType = 'ENSEMBL', 
                         ont = "BP")

# Plot upregulated GO terms
print("Upregulated genes GO enrichment:")
barplot(upGO_results, showCategory=10, title="GO Enrichment - Upregulated Genes")

# Get gene IDs for downregulated genes
downreg_genes <- rownames(downreg.df)
downGO_results <- enrichGO(gene = downreg_genes, 
                           OrgDb = "org.Hs.eg.db", 
                           keyType = 'ENSEMBL', 
                           ont = "BP")

# Plot downregulated GO terms
print("Downregulated genes GO enrichment:")
barplot(downGO_results, showCategory=10, title="GO Enrichment - Downregulated Genes")

# Using gProfiler for up and down regulated genes
go_upreg <- gost(upreg_genes, organism = 'hsapiens', significant = TRUE)
# gostplot doesn't accept a main/title parameter
p_upreg <- gostplot(go_upreg, interactive = TRUE, capped = TRUE)
print("GO enrichment plot for upregulated genes")
p_upreg

go_downreg <- gost(downreg_genes, organism = 'hsapiens', significant = TRUE)
p_downreg <- gostplot(go_downreg, interactive = TRUE, capped = TRUE)
print("GO enrichment plot for downregulated genes")
p_downreg
```
GSEA Plot, make one, Cesar will explain what it means 
```{R}
res <- res[order(-res$stat),]
gene_list <- res$stat
names(gene_list) <- rownames(res)
gse <- gseGO(gene_list, ont = "BP", keyType = "ENSEMBL", OrgDb ="org.Hs.eg.db", eps = 1e-300 )
gseaplot(gse, geneSetID = "GO:0048856" )

```

FINAL EXERCISE

Subset the 100 most expressed genes (in terms of logfold and p value) and generate a three GO plots base on BP, MF, CC. 
Select Print a GSEA figure of the most abundant GO from each of these classes
```{r}
# Subset the 100 most expressed genes (combining log fold change and p-value)
# First, remove genes with NA values
filtered_res <- sigs.df[!is.na(sigs.df$padj),]

# Create a score combining significance and fold change
filtered_res$score <- abs(filtered_res$log2FoldChange) * -log10(filtered_res$padj)

# Order by the combined score and take top 100
top_100 <- filtered_res %>% 
  arrange(desc(score)) %>% 
  head(100)

# Get the gene IDs for the top 100 genes
top_genes <- rownames(top_100)

# Run GO enrichment analysis for the three ontologies
# Add readable = TRUE parameter to avoid errors
go_BP <- enrichGO(gene = top_genes, 
                  OrgDb = "org.Hs.eg.db", 
                  keyType = 'ENSEMBL', 
                  ont = "BP",
                  readable = TRUE)

go_MF <- enrichGO(gene = top_genes, 
                  OrgDb = "org.Hs.eg.db", 
                  keyType = 'ENSEMBL', 
                  ont = "MF",
                  readable = TRUE)

go_CC <- enrichGO(gene = top_genes, 
                  OrgDb = "org.Hs.eg.db", 
                  keyType = 'ENSEMBL', 
                  ont = "CC",
                  readable = TRUE)

# Check if we have results before plotting
if (nrow(go_BP@result) > 0) {
  barplot(go_BP, showCategory=10, title="Biological Process GO Enrichment")
} else {
  print("No significant Biological Process terms found")
}

if (nrow(go_MF@result) > 0) {
  barplot(go_MF, showCategory=10, title="Molecular Function GO Enrichment")
} else {
  print("No significant Molecular Function terms found")
}

if (nrow(go_CC@result) > 0) {
  barplot(go_CC, showCategory=10, title="Cellular Component GO Enrichment")
} else {
  print("No significant Cellular Component terms found")
}

# Create a ranked gene list for GSEA (ordered by log2FoldChange)
ranked_genes <- res$stat
names(ranked_genes) <- rownames(res)
ranked_genes <- ranked_genes[!is.na(ranked_genes)]
ranked_genes <- sort(ranked_genes, decreasing = TRUE)

# Run GSEA for each ontology with error handling
# Use tryCatch to handle potential errors
tryCatch({
  gsea_BP <- gseGO(ranked_genes, 
                  ont = "BP", 
                  keyType = "ENSEMBL", 
                  OrgDb = "org.Hs.eg.db", 
                  minGSSize = 10,
                  maxGSSize = 500,
                  pvalueCutoff = 0.05)
  
  # Check if we have results
  if (nrow(gsea_BP@result) > 0) {
    top_BP_ID <- gsea_BP@result$ID[1]
    print("Top Biological Process GSEA Plot:")
    print(gseaplot(gsea_BP, geneSetID = top_BP_ID))
  } else {
    print("No significant BP terms found in GSEA")
  }
}, error = function(e) {
  print(paste("Error in BP GSEA:", e$message))
})

tryCatch({
  gsea_MF <- gseGO(ranked_genes, 
                  ont = "MF", 
                  keyType = "ENSEMBL", 
                  OrgDb = "org.Hs.eg.db", 
                  minGSSize = 10,
                  maxGSSize = 500,
                  pvalueCutoff = 0.05)
  
  if (nrow(gsea_MF@result) > 0) {
    top_MF_ID <- gsea_MF@result$ID[1]
    print("Top Molecular Function GSEA Plot:")
    print(gseaplot(gsea_MF, geneSetID = top_MF_ID))
  } else {
    print("No significant MF terms found in GSEA")
  }
}, error = function(e) {
  print(paste("Error in MF GSEA:", e$message))
})

tryCatch({
  gsea_CC <- gseGO(ranked_genes, 
                  ont = "CC", 
                  keyType = "ENSEMBL", 
                  OrgDb = "org.Hs.eg.db", 
                  minGSSize = 10,
                  maxGSSize = 500,
                  pvalueCutoff = 0.05)
  
  if (nrow(gsea_CC@result) > 0) {
    top_CC_ID <- gsea_CC@result$ID[1]
    print("Top Cellular Component GSEA Plot:")
    print(gseaplot(gsea_CC, geneSetID = top_CC_ID))
  } else {
    print("No significant CC terms found in GSEA")
  }
}, error = function(e) {
  print(paste("Error in CC GSEA:", e$message))
})
```

```{r}
gene_list <- diff.df$symbol
write.table(gene_list, file="gene_symbols.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```